File: Programming/02. CSharp Part 2/07.StringsTextProcessing/18.ExtractEmails/ExtractEmails.cs
//Write a program for extracting all email addresses from given text. 
//All substrings that match the format <identifier>@<host>…<domain> should be recognized as emails.

using System;
using System.Collections.Generic;
class ExtractEmails
{
    static void Main()
    {
        // text that has emails that are valid
        string text = @"Here are some emails that should be valid - " +
            "niceandsimple@example.com " +
            "very.common@example.com " +
            "a.little.lengthy.but.fine@dept.example.com " +
            "disposable.style.email.with+symbol@example.com " +
            "user@[IPv6:2001:db8:1ff::a0b:dbd0] " +
            @"""much.more unusual""@example.com " +
            @"""very.unusual.@.unusual.com""@example.com " +
            @"""very.(),:;<>[]\"".VERY.\""very@\\\""very\"".unusual""@strange.example.com " +
            "postbox@com (top-level domains are valid hostnames) " +
            "admin@mailserver1 (local domain name with no TLD) " +
            "!#$%&'*+-/=?^_`{}|~@example.org " +
            @"""()<>[]:,;@\\\""!#$%&'*+-/=?^_`{}|~.a""@example.org ";

        List<string> emailList = new List<string>();
        string[] emails = text.Split(' ');

        // regular expression pattern
        string pattern = @"[\w.!#$%&'*=?^_`{|}~\"" ""(),:;<>@[\]]{1,64}@[\w]{2,253}";
        //string pattern = @"([^\s""(),:;<>@[\]]{4,})\@(\w{3,})\.(\w{2,}(?:\.\w{2,})?)";

        foreach (string email in emails)
        {
            // if email that matchs the pattern is found
            if (System.Text.RegularExpressions.Regex.IsMatch(email, pattern))
            {
                // add it to the emailList
                emailList.Add(email);
            }
        }

        // display the emails
        foreach (var email in emailList)
        {
            Console.WriteLine(email);
        }
    }
}

File: HTML and CSS/JS/01.JS Fundamentals/11.Strings/js/extractEmails.js
// 9 Write a function for extracting all email addresses from given text. All substrings that match the format <identifier>@<host>…<domain> should be recognized as emails. Return the emails as array of strings.


function onExtractEnamils() {
    printInElement('', true);
    var text = "Some emails m.alexiev@abv.bg more mails malexiev@gmail.com Even more @mails that wont work like _ma@yahoo.com ! Just broken email@ malex@iev@gmail.com";

    printInElement(text + "\n");

    var emails = [];
    var index = text.indexOf("@");
    while (index > -1) {
        if (text[index - 1] != " " && text[index + 1] != " ") {
            var identifier = text.substring(text.lastIndexOf(" ", index - 1), index);
            var host = text.substring(index + 1, text.indexOf(".", index + 1));
            var domain = text.substring(text.indexOf(".", index + 1) + 1, text.indexOf(" ", text.indexOf(".", index + 1) + 1));
            var email = new String();

            // skip emails with more than one @ symbol
            if (identifier.indexOf("@") === -1 && host.indexOf("@") === -1 && domain.indexOf("@") === -1) {
                email = email.concat(identifier, "@", host, ".", domain);
                emails.push(email);
            }
        }
        index = text.indexOf("@", index + 1);
    }

    printInElement(emails);
}

